[MIG] stock_storage_type: Migration to 16.0

rousseldenis opened the pull request:
Depends on :

- [x] https://github.com/OCA/product-attribute/pull/1160
- [x] https://github.com/OCA/product-attribute/pull/1161
- [x] https://github.com/OCA/server-tools/pull/2413
- [x] https://github.com/OCA/stock-logistics-workflow/pull/1102
- [x] https://github.com/OCA/stock-logistics-workflow/pull/1103
- [x] https://github.com/OCA/stock-logistics-workflow/pull/1104
- [x] https://github.com/OCA/stock-logistics-warehouse/pull/1509
- [x] https://github.com/OCA/stock-logistics-warehouse/pull/1655
- [x] https://github.com/OCA/stock-logistics-workflow/pull/1274

TODO: 

- [x] Remove test-requirements.txt entries

grindtildeath committed:
[ADD] stock_storage_type

oca-travis committed:
[UPD] Update stock_storage_type.pot

grindtildeath committed:
Move restrictions from stock_putaway_storage_type_strategy to base module

Define max height/max weight restriction fields

Dedicated tree view to edit package/location storage type relation

Rename fields to drop stock_ prefix

grindtildeath committed:
[IMP] stock_storage_type: black, isort

grindtildeath committed:
[MIG] stock_storage_type: Migration to 13.0

Add tests

Fix dependency issue for pytest

[IMP] pre-commit XML files

Add setuptools

Add module stock_storage_type_putaway_strategy

[IMP] stock_storage_type_putaway_strategy: black, isort

[MIG] stock_storage_type_putaway_strategy: Migration to 13.0

Compute domain on stock.package_level location_dest_id according to storage_type restrictions

Allow to compute restrictions for multiple products and quants

Add constraint on stock.quant according to stock_storage_type

Add tests

Add tests on package destination domain

Add package_level actual location to allowed domain

Fix constraints do not mix products/lots

grindtildeath committed:
Merge putaway_strategy module into stock_storage_type

Fix failing test and add loggers

Improve error message on stock.quant constraint

Commit:
Apply pre-commit

Check waiting moves in "Not Empty" put-away constraints

When an unreserved move's destination is already a bin, we assume:

* the move is probably waiting on another that will be processed soon
* as the destination of the move is already a bin (not Stock, ...), the
  good *will* arrive in this precise location and not another

So we prevent new move lines to target this location.

Note: as we do not know the lot yet in the move, we cannot check if we
have a planned lot for the same lot in the "Do not mix same lot": ignore
it.

gurneyalex committed:
[IMP] stock_storage_type

improve perfs
* in _compute_allowed_location_dest_ids handle the case of a missing pack level, and short cut
* move a mapped() call out of a loop
* use the stored related field state of stock.move.line instead of a dotted path in the search domain

gurneyalex committed:
Improve put-away location selection perfs

The choice of a suitable stock location was implemented using an
iteration on all candidate locations + several filters running for each.
This is not efficient when we have several 1000s of locations (each
check was taking 30ms -> several minutes required to compute a putaway
location). This commit reimplements the algorithm by using queries to
select locations matching the rules. In order to be able to compute this
using the ORM, we added some technical fields which should not harm
performance too much.

Co-authored-by: Guewen Baconnier <guewen.baconnier@camptocamp.com>
Co-authored-by: Akim Juillerat <akim.juillerat@camptocamp.com>

Commit:
Find leaf sub-locations instead of all children

We will never do a put-away in an intermediate location,
so they should never be kept as candidate locations for the
put-away.

Commit:
Replace mapped by a SQL query

Which is much smaller with thousands of locations for a couple of
storage types.

I recycled the field location_ids on the location storage type which
was a unused field (probably forgotten during a refactoring).

Commit:
Use a pre-generated domain to filter allowed locations

When we select a package in a package level, we want to restrict
the selection of the location destination based on the put-away rules.

The previous implementation used a computed M2m field, added in the
view as an invisible field, and a domain on the location_dest_id field
filtering the locations based on this m2m.

The issue: with 5000 allowed locations, changing the package triggers
an onchange (that we want), and the server answers with 5000 ids,
which must be processed by the M2m field client side. It blocks
the browser for several seconds.

The module web_domain_field [0] allow to use a domain stored in a Char
field rather than a field in the XML (read the module's description if
not clear). Using it solves this issue, as the client doesn't need to
evaluate and update a m2m field, it only pass the domain downstream.

[0] https://github.com/OCA/web/tree/13.0/web_domain_field

Commit:
Remove constrains forbidding 'none' + storage type on location

It may be a valid use case, for instance to be able to send the good on
the "zone" above the bins (with a following new move to go from the Zone
to the bin).

grindtildeath committed:
pre-commit

oca-travis committed:
[UPD] Update stock_storage_type.pot

OCA-git-bot committed:
[UPD] README.rst

OCA-git-bot committed:
stock_storage_type 13.0.1.0.1

Commit:
Allow to select locations with 'none' strategy in putaway sequence

The putaway sequence for a package storage types defines where the goods
can be put and in which order.

For instance, if the "cardbox" storage type has this sequence:

1. Stock/Cardboxes Shelving: ordered locations
2. Stock/Cardboxes Reserve: ordered locations

It will first try to put the goods in Bin 1, then Bin 2 and finally Bin
3 if the others are full.

This commit aims to support the strategies with a 'none' strategy has
valid locations in the sequences, such as:

1. Stock: None
2. Stock/Cardboxes Shelving: ordered locations
3. Stock/Cardboxes Reserve: ordered locations

In this example, if a move line with a cardbox has a destination Stock,
no strategy will be applied, if the move line has a destination Cardbox
Shelving, the ordered location strategy is applied.

It allows to use a sort of buffer: the default destination is Stock,
then the destination is updated to Stock/Cardbox Reserve (upon the
choice of a user!) and a new computation of the putaway is asked.

Commit:
Ignore empty quants in constraints

oca-travis committed:
[UPD] Update stock_storage_type.pot

OCA-git-bot committed:
[UPD] README.rst

OCA-git-bot committed:
stock_storage_type 13.0.1.1.0

Commit:
Add stock_storage_type_buffer

This is an extensions for putaway with storage types (provided by
``stock_storage_type``).

In a warehouse, some areas where goods are put cannot be accessed
directly and need to be moved to a buffer first (for instance, because
they need a special machine to move from the buffer to the bins).

This module allows to define buffers in front of a set of locations.
When a buffer contains something or has move lines that reaches it, all
the locations are removed from the put-away computation.

grindtildeath committed:
stock_storage_type: Add package storage type on product

Package storage type can be defined on products to compute automatically
a package storage type on packages if no packaging is defined and
the package only contains a single product.

Commit:
Make stock_storage_type depend on stock_putaway_hook

The putaway based on storage type *must* be applied at the very end
of the putaway rules, because it depends on the putaway location found.

If we execute it by calling super() and we added different putaway
strategies in StockLocation, we have no way to be sure it is the last
called.

That's why now it depends on stock_putaway_hook, which allow to search
for alternative put-away hooks, and call _putaway_strategy_finalizer at
the very end, where we can plug the storage type rules.

grindtildeath committed:
Update stock_storage_type/models/stock_quant_package.py

Co-authored-by: Guewen Baconnier <guewen.baconnier@camptocamp.com>

OCA-git-bot committed:
stock_storage_type 13.0.1.2.0

oca-travis committed:
[UPD] Update stock_storage_type.pot

OCA-git-bot committed:
stock_storage_type 13.0.1.3.0

sebalix committed:
stock_storage_type: option to make 'height' required on packages

sebalix committed:
[IMP] stock_storage_type: sort putaway locations by supported max_height

New max_height field on the the location storage type to indicate the
max height supported by a location among all its storage type.

We want to propose the first location corresponding to the height of
a package by losing a minimum of space, so we sort leaf locations on
their supported max height.

oca-travis committed:
[UPD] Update stock_storage_type.pot

OCA-git-bot committed:
stock_storage_type 13.0.1.4.0

sebalix committed:
[IMP] stock_storage_type_putaway_abc: sort also locations by max heights

For ABC strategy we apply the same sort among on each chunk of
locations "A", "B" and "C".

If we have the following locations:

    LOC_A_MAXHEIGHT_200
    LOC_A_MAXHEIGHT_100
    LOC_B_MAXHEIGHT_400
    LOC_B_MAXHEIGHT_300
    LOC_C_MAXHEIGHT_600
    LOC_C_MAXHEIGHT_500

If the ABC storage of a product is flagged B, then the proposed
locations will be ordered as:

    LOC_B_MAXHEIGHT_300
    LOC_B_MAXHEIGHT_400
    LOC_C_MAXHEIGHT_500
    LOC_C_MAXHEIGHT_600
    LOC_A_MAXHEIGHT_100
    LOC_A_MAXHEIGHT_200

(ordered at first with BCA as usual, then by max_height)

Commit:
Fix unpredictable package storage type on packages

From the information I collected, the requirements for the
``package_storage_type_id`` field on the packages are:

* if we have a package storage type on the related product packaging,
  set it on the package
* else if we have only one product in the package and the product has a
  package storage type, set it on the package
* users must be able to change the storage type, so we cannot keep it
  in sync as a related field would do

The solution using a stored computed field which is set only once is not
reliable: if the computation of the field is called before the product
packaging is set (by the auto assign method), and we have a single
product in the package, the product's storage type is set. We expect to
have the packaging's storage type in this case.

Instead, use a normal non-computed field and set it from:

* when creating a new package with a product packaging or writing a
  product packaging, overwrite the storage type with the new one
* the auto_assign_packaging method sets the "single product" storage
  type as fallback when there is no product packaging on the package

Commit:
Fix sorting broken by a limit on search for valid locations

In this context, "self" are sorted locations. The search returns valid
locations. Later, we filter the locations in "self" by keeping only the
valid locations, but keeping the order of "self.ids".  If we apply a
limit on the intermediate search for valid locations, the order of
"self.ids" may be wrong as the search will return only the first valid
id(s) according to its own order which may be different.

Commit:
Fix manager access on stock storage types

All users have group_adv_location or group_stock_multi_locations when the options are activated.
Normal users should not be allowed to edit these models, only actual managers

OCA-git-bot committed:
stock_storage_type 13.0.1.4.1

oca-travis committed:
[UPD] Update stock_storage_type.pot

OCA-git-bot committed:
stock_storage_type 13.0.1.5.0

Commit:
Extract methods in stock_storage_type to allow extensions

Commit:
Add small optimization on sorted intersection

Commit:
Add outgoing move lines qty_done qty in "is_empty"

A location is considered empty if all the quantity of its quants
are already moved by the "qty_done" of the outgoing move lines.

oca-travis committed:
[UPD] Update stock_storage_type.pot

OCA-git-bot committed:
stock_storage_type 13.0.1.6.0

Commit:
stock_storage_type: fix location w/ no leaf locations

max() cannot be called on an empty iterable, as we have no locations
anyway, return the empty recordset

sebastienbeau committed:
[IMP] black, isort, prettier

sebastienbeau committed:
[IMP] set module uninstallable

simahawk committed:
stock_storage_type: fix warning non writeable field

'single_product_id' is not writeable and cannot be used in constraints.

hparfr committed:
[MIG] stock_storage_type: Migration to 14.0

sebalix committed:
stock_storage_type: use estimated package weight if the weight is not set

Commit:
Optimize "will contain" computed fields on stock.location

The 2 fields:

* location_will_contain_product_ids
* location_will_contain_lot_ids

Are used only when applying a related location storage type with
"do_not_mix_products" or "do_not_mix_lots". As the result of the
computed fields are stored in m2m related fields, we can prevent a
lot of insertions in these relation tables by not computing them
if they have no location storage type that will need them.

Commit:
Add barcode on Package Storage Type

Commit:
Exclude non-internal locations for storage types computed fields

We'll never do put-away in special locations (supplier, customer, ...),
only in internal locations. We can then ignore computation for the "will
contain product/lot" fields and for the is empty field for special
locations.

Commit:
stock_storage_type: prevent useless writes on stock.location

Complement previous commits: e89b2819 and 461a26e0
We want to prevent writing these computed fields when unnecessary,
for instance, on non internal locations, we should not update anything
if the value was already correct.

However, when the value has to be computed (internal location, ...),
we want to write the value if it didn't change, so concurrent
transactions are retried.

An example: if we have a quant of 20 in a location, and 2 concurrent
transaction each take a quant,

0. Initial value for "location_is_empty": False
1. Tx A: set a qty done on move line for taking 12
2. Tx B: set a qty done on move line for taking 8

None of the transactions see that the other took the rest, none of them
change the `location_is_empty` to True. But if they both try to write
`location_is_empty: False` on the location, one of them is rejected and
retry after the other is committed: this time the second one see it
takes the last 8 and change `location_is_empty` to True.

Commit:
Optimize location_is_empty to be computed only when necessary

Follow-up of: e89b2819, 461a26e0, and 9a3da83d

The same way the fields "location_will_contain_product_ids" and
"location_will_contain_lot_ids" are computed only when the stock
location has a location storage type "do not mix products" or "do not
mix lots", the "location_is_empty" field is computed only when there
is a storage type with "Only Empty".

The module "stock_storage_type_buffer" also needs no know if a location
is empty following the same rules, so buffers are added to the locations
that need "location_is_empty" to be computed.

Important: the fact that the computed fields write on "only_empty" even
if the  value does not change is on purpose: it prevents 2 concurrent
transactions to pick or put in the same location at the same time, which
could have an effect on this field without knowing about the other move.
(eg. field is "not empty" and 2 moves pick half of the goods, if we have
no error and retry of one of the transaction, the location will remain
"not empty" without actual quantity).

A related improvement that could yet be done, is to lock the location
when we pick/put so other transactions that list all the available
put-aways (e.g. in the m2o widget of a package level) can remove them.
It could prevent a retry of the transaction if before the concurrent
failure we could select "the next" available location"

Commit:
Improve readability by using proper variable name

The records we iterate on are not locations but storage types.

lmignon committed:
stock_storage_type: putaway strategy "do not mix" should give priority to location that already contains the product (the one with the less qty) fixes #168

lmignon committed:
stock_storage_type: Optimisation of the sorting of valid locations

Co-authored-by: Jacques-Etienne Baudoux <je@bcim.be>

lmignon committed:
stock_storage_type: Finalisation of the optimisation of the sorting of valid locations

lmignon committed:
stock_storage_type: browse only once and avoid recordset union

sebalix committed:
s_storage_type: no storage type for delivery

There is no need to set a package storage type on delivery packages (we
are not storing them).
This avoid to trigger blocking constraints (like height).

Commit:
stock_storage_type: allow to archive storage types/locations

* Usual field added.
* Simple tests added, to keep the coverage.
* Filter added to show the archived records.

jbaudoux committed:
Change sorting of locations in the pack putaway computation

Invert the sorting for abc packs storage strategy to first apply the
max_height which is a physical constraint, then apply the abc
preference.

Add a pack putaway sequence to be able to favor some locations (e.g.
level). For ordered children strategy, finalize by sorting by location
name. For abc, random selection is applied at last as it was already.

Replaced multi-steps python sorting by an unique SQL quicksort query to
quickly compute the result on large stock_location table. This is
especially important for operations containing many moves requiring a
putaway computation.

jbaudoux committed:
Fix tests for stock_vertical_lift_storage_type by ensuring order when tray cells have all the same name

hparfr committed:
stock_storage_type: fix fields changed in dependencies

weight and height have now explicit uom

oca-travis committed:
[UPD] Update stock_storage_type.pot

OCA-git-bot committed:
[UPD] README.rst

sebalix committed:
[IMP] stock_storage_type: black, isort, prettier

oca-travis committed:
[UPD] Update stock_storage_type.pot

OCA-git-bot committed:
stock_storage_type 14.0.1.1.0

oca-travis committed:
[UPD] Update stock_storage_type.pot

OCA-git-bot committed:
stock_storage_type 14.0.1.2.0

OCA-git-bot committed:
stock_storage_type 14.0.1.3.0

OCA-git-bot committed:
stock_storage_type 14.0.1.3.1

sebalix committed:
[FIX] s_storage_type: Search tag requires at least one field element

OCA-git-bot committed:
stock_storage_type 14.0.1.4.0

oca-travis committed:
[UPD] Update stock_storage_type.pot

OCA-git-bot committed:
stock_storage_type 14.0.1.5.0

sebalix committed:
[IMP] stock_storage_type: black, isort, prettier

oca-travis committed:
[UPD] Update stock_storage_type.pot

OCA-git-bot committed:
stock_storage_type 14.0.1.6.0

oca-travis committed:
[UPD] Update stock_storage_type.pot

OCA-git-bot committed:
stock_storage_type 14.0.1.7.0

jbaudoux committed:
Fix Package storage type message computation.

The message was broken with the changes in commit 6b25b867c2f606d5f4220915ab9aae3b30c1f87f

jbaudoux committed:
stock_storage_type: add location smart buttons

oca-travis committed:
[UPD] Update stock_storage_type.pot

OCA-git-bot committed:
[UPD] README.rst

OCA-git-bot committed:
stock_storage_type 14.0.1.8.0

jbaudoux committed:
[FIX] stock_storage_type: Fix putaway compution when a strategy is defined

When the strategy has to find a leaf sub-location, do not return the
location itself as defined on a storage type putaway sequence.

Example: sequence is defined to first look in WH/Stock/area1 and then in
WH/Stock/area2. On area1 the strategy is to find a leaf sub-location
(strategy = Ordered Children Location) under area1. If area1 has
temporarily no sub-location, area1 should not be returned as a valid
leaf location.

Also, when searching for a leaf location, exclude non active locations

sebalix committed:
[FIX] s_storage_type: allowed location storage types

Rewrite the `_compute_allowed_location_storage_type_ids` in a iterative
way (instead of using recursion) to avoid a RecursionError exception
when computing large recordsets.

OCA-git-bot committed:
stock_storage_type 14.0.1.8.1

sebalix committed:
Revert "[FIX] s_storage_type: allowed location storage types"

This reverts commit b35b9d53525f1777b4e55e0356707f9388c13031.

OCA-git-bot committed:
stock_storage_type 14.0.1.8.2

ibuioli committed:
Added translation using Weblate (Spanish (Argentina))

oca-travis committed:
[UPD] Update stock_storage_type.pot

OCA-git-bot committed:
stock_storage_type 14.0.1.9.0

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: wms-14.0/wms-14.0-stock_storage_type
Translate-URL: https://translation.odoo-community.org/projects/wms-14-0/wms-14-0-stock_storage_type/

ibuioli committed:
Translated using Weblate (Spanish (Argentina))

Currently translated at 35.9% (41 of 114 strings)

Translation: wms-14.0/wms-14.0-stock_storage_type
Translate-URL: https://translation.odoo-community.org/projects/wms-14-0/wms-14-0-stock_storage_type/es_AR/

ibuioli committed:
Translated using Weblate (Spanish (Argentina))

Currently translated at 71.9% (82 of 114 strings)

Translation: wms-14.0/wms-14.0-stock_storage_type
Translate-URL: https://translation.odoo-community.org/projects/wms-14-0/wms-14-0-stock_storage_type/es_AR/

ibuioli committed:
Translated using Weblate (Spanish (Argentina))

Currently translated at 100.0% (114 of 114 strings)

Translation: wms-14.0/wms-14.0-stock_storage_type
Translate-URL: https://translation.odoo-community.org/projects/wms-14-0/wms-14-0-stock_storage_type/es_AR/

lmignon committed:
[IMP] stock_storage_type: Allow to restrict some putaway sequence

fixes #343

lmignon committed:
[IMP] stock_storage_type: Remove useless code

The pack_putaway_strategy==None and the validation of the sequence location is also done into the code of the method 'get_storage_locations'. This code could therefore be removed with the benefit of processing the strategy into the same place

lmignon committed:
[IMP] stock_storage_type: Add support for fixed location put away

If the put-away strategy is None, we also need to compute the odoo standard put-away on that destination.

refs #343

simahawk committed:
stock_storage_type: dev status -> Beta

OCA-git-bot committed:
[UPD] README.rst

OCA-git-bot committed:
stock_storage_type 14.0.1.10.0

OCA-git-bot committed:
stock_storage_type 14.0.1.11.0

sebalix committed:
Fix access right issues when using actions

OCA-git-bot committed:
stock_storage_type 14.0.1.11.1

sebalix committed:
[FIX] s_storage_type: fix computation of 'max_height_in_m'

The computation of 'max_height_in_m' depends also on 'length_uom_id'
field.

OCA-git-bot committed:
stock_storage_type 14.0.1.11.2

Commit:
[UPD] Update stock_storage_type.pot

OCA-git-bot committed:
[UPD] README.rst

OCA-git-bot committed:
stock_storage_type 14.0.1.12.0

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: wms-14.0/wms-14.0-stock_storage_type
Translate-URL: https://translation.odoo-community.org/projects/wms-14-0/wms-14-0-stock_storage_type/

ibuioli committed:
Translated using Weblate (Spanish (Argentina))

Currently translated at 100.0% (125 of 125 strings)

Translation: wms-14.0/wms-14.0-stock_storage_type
Translate-URL: https://translation.odoo-community.org/projects/wms-14-0/wms-14-0-stock_storage_type/es_AR/

flachica committed:
[MIG] stock_storage_type: 15.0

rousseldenis committed:
[IMP] stock_storage_type: black, isort, prettier

rousseldenis committed:
[MIG] stock_storage_type: Migration to 16.0

rousseldenis committed:
[IMP] stock_storage_type: Remove web_domain_field module dependency

rousseldenis committed:
[IMP] stock_storage_type: Change group from view to field

rousseldenis committed:
[CHG] stock_storage_type: Use of core stock.package.type object

As <stock.location>._apply_putaway_strategy() is using the package_type_id field
to pass the stock.quant.package as parameter, we don't need anymore the
quant passed through the context and we override that core model by adding
fields we used in former stock.package.storage.type model.

rousseldenis commented:
/ocabot migration  stock_storage_type

rousseldenis commented:
@jbaudoux @lmignon I think this is a first testable version.

rousseldenis committed:
[IMP] stock_storage_type: Add dependency on stock_package_type_button_box

rousseldenis commented on stock_storage_type/tests/test_storage_type_move.py:
```diff
@@ -346,29 +352,29 @@ def _levels_for(packages):
         # Cardbox with different product go into different cardbox location
         self.assertEqual(len(third_level), 1)
-        self.assertEqual(third_level.location_dest_id, self.cardboxes_bin_2_location)
+        self.assertEqual(third_level.location_dest_id, self.cardboxes_bin_3_location)
```
@jbaudoux @lmignon This is the only test case that I've changed.

It seems that order has been changed

jbaudoux commented:
@rousseldenis Do you think we can replace `stock.location.storage.type` by the new standard `stock.storage.category` and add a tab on the "Storage Category" called "Capacity by Storage" where you show the list of available storage types?
Drawback I see is:
* the max_height won't be supported for the capacity by package type or product
* the location can only have a single storage category but that's not bad limitation

rousseldenis commented:
> @rousseldenis Do you think we can replace `stock.location.storage.type` by the new standard `stock.storage.category` and add a tab on the "Storage Category" called "Capacity by Storage" where you show the list of available storage types? Drawback I see is:
> 
> * the max_height won't be supported for the capacity by package type or product
> * the location can only have a single storage category but that's not bad limitation

Not yet tested but that can make sense to get back to standard.

jbaudoux commented:
@rousseldenis Coming back on this:
> * the location can only have a single storage category but that's not bad limitation

I checked more deeply the implications and that would mean that the `allow_new_product` needs to be also added on the relation between the storage category and the product/pack storage. It would be an optional setting overriding the default one from the category.

Today we are using this config: Location A has:
* location storage type LST-1 is empty for product/pack storage type PST-X
* location storage type LST-2 is do not mix for product/pack storage type PST-Y

That would become: location A has:
* location storage category LC-12 for product/pack storage type PST-X in mode is empty and for PST-Y in mode do not mix

lmignon commented:
> @rousseldenis Coming back on this:
> 
> > * the location can only have a single storage category but that's not bad limitation
> 
> I checked more deeply the implications and that would mean that the `allow_new_product` needs to be also added on the relation between the storage category and the product/pack storage. It would be an optional setting overriding the default one from the category.
> 
> Today we are using this config: Location A has:
> 
>     * location storage type LST-1 is empty for product/pack storage type PST-X
> 
>     * location storage type LST-2 is do not mix for product/pack storage type PST-Y
> 
> 
> That would become: location A has:
> 
>     * location storage category LC-12 for product/pack storage type PST-X in mode is empty and for PST-Y in mode do not mix

If we add information on the relation between storage category and pack/product storage, we could also add the max_weight at the same level

jbaudoux commented:
> If we add information on the relation between storage category and pack/product storage, we could also add the max_weight at the same level

The max weight/height is a location characteristic not depending on what you put in. According to me, no need to put that on the relation.

lmignon commented:
> The max weight/height is a location characteristic not depending on what you put in. According to me, no need to put that on the relation.

It makes sens. That means we can remove all the logic used to compute this max_height on the stock_location and declare the field on the stock_location if it doesn't exist.

lmignon commented:
@jbaudoux 

Since `stock.package.storage.type`is now a `stock.package.type`, we already have the relation on `stock.storage.category`. The only missing piece according to our last comment is that we should add an `allow_new_product` optional field on this relation. (Implemented by `stock.storage.category.capacity`)

jbaudoux commented:
> Since `stock.package.storage.type`is now a `stock.package.type`, we already have the relation on `stock.storage.category`. The only missing piece according to our last comment is that we should add an `allow_new_product` optional field on this relation. (Implemented by `stock.storage.category.capacity`)

No, you cannot replace `stock.package.storage.type`by `stock.package.type`. It is a complete different concept. The package type will say for example that it is a pallet. The storage type will tell the putaway configuration to apply. We need the 2 concepts as 2 different products on a pallet can require completely different putaway configuration.

jbaudoux commented:
> It makes sens. That means we can remove all the logic used to compute this max_height on the stock_location and declare the field on the stock_location if it doesn't exist.

It can be replaced by a related readonly to the storage category.

rousseldenis commented:
> No, you cannot replace `stock.package.storage.type`by `stock.package.type`. It is a complete different concept.

@jbaudoux Have you tested it on runboat ? I made it working with nothing changed at tests levels.

jbaudoux commented:
> > No, you cannot replace `stock.package.storage.type`by `stock.package.type`. It is a complete different concept.
> 
> @jbaudoux Have you tested it on runboat ? I made it working with nothing changed at tests levels.

The issue is that it is impossible to configure because it is 2 different concepts. We need multiple package storage types for a same package type. The package storage type represent the putaway rule you set for the package type. And we have many rules (package storage type) for a box or a pallet (package type).

jbaudoux reviewed (changes requested):
You cannot replace stock.package.storage.type by stock.package.type.
See previous comments

jbaudoux commented:
> The issue is that it is impossible to configure because it is 2 different concepts. We need multiple package storage types for a same package type. The package storage type represent the putaway rule you set for the package type. And we have many rules (package storage type) for a box or a pallet (package type).

After discussion, we can use the new `stock.package.type` instead of `stock.package.storage`. However, this is not to be confused by the package type OCA module (https://github.com/OCA/product-attribute/tree/14.0/product_packaging_type) which has then to not be replaced by the new `stock.package.type`

rousseldenis commented:
@jbaudoux I try to finish the refactor and I have one question about the combinations we could have on a particular location about strategies (do not mix products / mix - empty/not empty - same lot/not same lot).

Here, the first approach I took is to have a related field 'allow_new_product' on locations that comes from the storage category.

The problem is that I loose the values that are on capacities.

Do we want to keep that flexibility on strategies possible combinations ?

Like determined by an 'any()' : https://github.com/OCA/wms/blob/14.0/stock_storage_type/models/stock_location.py#L148

As we have now the information on two levels (category and capacity).

So, to summarize:

- Either we only have the 'alllow_new_product' value coming from the category
- Either we keep all 'do_not_....' fields on locations computed on (category or/and capacities)


It"s quite strategic as this is the way locations are filtered in order to determine the first good one.

Thanks for your input

jbaudoux commented:
This one:
> * Either we keep all 'do_not_....' fields on locations computed on (category or/and capacities)

rousseldenis commented:
> This one:
> 
> > * Either we keep all 'do_not_....' fields on locations computed on (category or/and capacities)

@jbaudoux And should it be computed only on capacities or on both capacities and category values (for each allow_new_product) ?

rousseldenis commented:
> This one:
> 
> > * Either we keep all 'do_not_....' fields on locations computed on (category or/and capacities)

And last but not least:

- As we have a 1 -> 1 relation between location > category and as capacities belong to one and only one category, we can have only one value of the 'allow_new_product' coming from the capacities. So, the fields computed on location side can have just one combination at one time. So, should we keep instead the 'do not mix lots' that I've added to 'allow_new_product' possible values on capacities side ?

jbaudoux commented:
> > This one:
> > > * Either we keep all 'do_not_....' fields on locations computed on (category or/and capacities)
> 
> And last but not least:
> 
> * As we have a 1 -> 1 relation between location > category and as capacities belong to one and only one category, we can have only one value of the 'allow_new_product' coming from the capacities. So, the fields computed on location side can have just one combination at one time. So, should we keep instead the 'do not mix lots' that I've added to 'allow_new_product' possible values on capacities side ?

On the category & capacity, we need to add that option "do not mix lots".
Then on the location, it serves only to know what to compute further. So you probably need to keep all the booleans on the location and compute them based on category and capacity. True if any is True. However, the boolean could be moved to the category.

rousseldenis commented:
> > > This one:
> > > > * Either we keep all 'do_not_....' fields on locations computed on (category or/and capacities)
> > 
> > 
> > And last but not least:
> > 
> > * As we have a 1 -> 1 relation between location > category and as capacities belong to one and only one category, we can have only one value of the 'allow_new_product' coming from the capacities. So, the fields computed on location side can have just one combination at one time. So, should we keep instead the 'do not mix lots' that I've added to 'allow_new_product' possible values on capacities side ?
> 
> On the category & capacity, we need to add that option "do not mix lots". Then on the location, it serves only to know what to compute further. So you probably need to keep all the booleans on the location and compute them based on category and capacity. True if any is True. However, the boolean could be moved to the category.

Geat! Let's go

rousseldenis committed:
[IMP] stock_storage_type: Move stock.location.storage.type to core capacities

As Odoo introduced Storage Categories and Storage Capacities, we'll use them in order
to represent the location restrictions.

rousseldenis commented on stock_storage_type/models/stock_location.py:
```diff
@@ -0,0 +1,652 @@
+        "computed_storage_category_id.do_not_mix_lots",
+        "computed_storage_category_id.capacity_ids.do_not_mix_lots",
+    )
+    def _compute_do_not_mix_lots(self):
```
@jbaudoux So, the last thing is to move this to storage category.

As there is a manual field there too. Do you see an approach to allow setting a value on the category itself apart the capacities (colliding field names) ?

rousseldenis replied:
@jbaudoux

jbaudoux replied:
resolved

jbaudoux commented on stock_storage_type/models/stock_storage_category_capacity.py:
```diff
@@ -0,0 +1,149 @@
+        default="mixed",
+        required=True,
+    )
+    do_not_mix_lots = fields.Boolean(
```
This needs to be replaced by an additional option in the selection `allow_new_product` that you can name `same_lot` = "If all lots are the same"

rousseldenis replied:
@jbaudoux I tried to do it but I don't remember the case where this does not work. I'll keep you in touch

jbaudoux commented on stock_storage_type/views/stock_storage_category.xml:
```diff
@@ -0,0 +1,21 @@
+        <field name="arch" type="xml">
+            <field name="package_type_id" position="after">
+                <field name="allow_new_product" />
+                <field
```
include as an option of `allow_new_product`

jbaudoux commented on stock_storage_type/models/stock_quant.py:
```diff
@@ -0,0 +1,155 @@
+                        ).format(storage_capacity=capacity.display_name)
+                    )
+                    continue
+                if capacity.do_not_mix_lots and (
```
```suggestion
                if capacity.allow_new_product == "same_lot" and (
```

jbaudoux commented on stock_storage_type/models/stock_storage_category.py:
```diff
@@ -0,0 +1,94 @@
+        comodel_name="stock.location", inverse_name="computed_storage_category_id"
+    )
+
+    do_not_mix_lots = fields.Boolean(
```
This needs to be replaced by an additional option in the selection `allow_new_product` that you can name `same_lot` = "If all lots are the same"

jbaudoux commented on stock_storage_type/models/stock_storage_category_capacity.py:
```diff
@@ -0,0 +1,149 @@
+        ]
+        # TODO this method and domain is applied once per storage type. If it's
+        # too slow at some point, we could group the storage types by similar
+        # configuration (only_empty, do_not_mix_products, do_not_mix_lots) and
```
Comment to update with new selection options

rousseldenis committed:
[IMP] stock_storage_type: Add migration scripts

rousseldenis committed:
[IMP] stock_storage_type: Rename stock.production.lot to stock.lot

rousseldenis committed:
[IMP] stock_storage_type: Add recursive property on fields that depends on self

rousseldenis committed:
[IMP] stock_storage_type: Merge 'do not mix lots' strategy into 'allow_new_product' one

rousseldenis committed:
[IMP] stock_storage_type: Migration scripts

rousseldenis commented:
@jbaudoux @lmignon Could you do a review pass on this one ?

lmignon commented on stock_storage_type/models/stock_location.py:
```diff
@@ -0,0 +1,653 @@
+            _logger.debug(
+                "Computing putaway for pack %s (%s)"
+                % (quants.package_id, quants.package_id)
+            )
```
@rousseldenis When no quant is provided you should look on the product to get the `package_type_id` . see https://github.com/acsone/wms/commit/48454dcb6731fc2c4bc805a882a82713e1a93ad4

rousseldenis replied:
@lmignon This has been solved there https://github.com/acsone/wms/commit/48454dcb6731fc2c4bc805a882a82713e1a93ad4

lmignon reviewed (changes requested):
Thank you @rousseldenis. A little comment...

rousseldenis committed:
[IMP] stock_storage_type: Use default product package type

When the product to move is not contained in a package, use
the default package type from the product if it is set.

rousseldenis commented:
@lmignon @jbaudoux I think all comments have been attended.

lmignon reviewed (commented):
@rousseldenis At first read, all this looks good to me. We need to make more tests to ensure that everything is working as expected on our database.

rousseldenis commented on stock_storage_type/models/stock_location.py:
```diff
@@ -0,0 +1,661 @@
+        )
+        return putaway_location
+
+    def get_storage_locations(self, products=None):
```
@jbaudoux We are questionning the 'products' attribute as there is no need of 'multiple' products anywhere.

The code calling this method is (here above):


![image](https://user-images.githubusercontent.com/19529533/203337569-944d803f-60a8-4d54-b4bc-dc19e4601b08.png)





We tend to say, we should change to 'product'.

If yes we could also remove the first() method there:

https://github.com/OCA/wms/blob/14.0/stock_storage_type_putaway_abc/models/stock_location.py#L40

But maybe you have a case where this is a recordset

rousseldenis replied:
@jbaudoux

jbaudoux replied:
This is required because a package can contain multiple products. This was clarified by adding a "s" to the product argument of the method call.
Can be marked as resolved.

sbejaoui reviewed (approved):
LGTM (code review)

rousseldenis commented:
@jbaudoux @lmignon

simahawk commented:
There's plenty of changes here... Do you mind if we add the module as not installable in 16.0 to ease reviews?

lmignon commented:
> There's plenty of changes here... Do you mind if we add the module as not installable in 16.0 to ease reviews?

`not installable in 16.0 to ease reviews`.???? How can you review a module if you're not able to install-it to make functional tests?
For sure there are plenty of changes since some concepts into the initial implementation are now part of the odoo standard. The migration has been a complex process to partially reconcile the initial implementation with the odoo standard without loosing functionalities.  We use this branch in our deployment and it should remain installable to not break it.

simahawk commented:
@lmignon I have different opinion regarding the review, but ok. Maybe we can open another one... or I surrender and I can review commit by commit :)

simahawk commented:
>  How can you review a module if you're not able to install-it to make functional tests?

BTW I don't get this. Your PR here will still be installable and you can run tests and do all you want as normal...

jbaudoux commented:
> There's plenty of changes here... Do you mind if we add the module as not installable in 16.0 to ease reviews?

The adaptation to v16 had been finalized by Denis with required refactoring. I'm going to retest it and update my review.
This is about to be approved

jbaudoux commented on stock_storage_type/models/stock_storage_location_sequence.py:
```diff
@@ -0,0 +1,91 @@
+            (
+                "allowed_location_storage_type_ids",
+                "in",
+                self.package_type_id.location_storage_type_ids.ids,
```
AttributeError: 'stock.package.type' object has no attribute 'location_storage_type_ids'

jbaudoux commented on stock_storage_type/models/stock_location.py:
```diff
@@ -0,0 +1,661 @@
+    def _compute_computed_storage_category_id(self):
+        """
+        This computes the Storage Storage depending on:
+          - its own Storage Storage
```
```suggestion
        This computes the Storage Category depending on:
          - its own Storage Category
```

jbaudoux commented on stock_storage_type/views/stock_location.xml:
```diff
@@ -0,0 +1,22 @@
+        <field name="inherit_id" ref="stock.view_location_form" />
+        <field name="arch" type="xml">
+            <field name="storage_category_id" position="after">
+                <field name="computed_storage_category_id" readonly="1" />
```
```suggestion
                <field name="computed_storage_category_id" readonly="1" groups="stock.group_stock_storage_categories"/>
```

jbaudoux commented on stock_storage_type/views/stock_location.xml:
```diff
@@ -0,0 +1,22 @@
+                    name="computed_storage_capacity_ids"
+                    widget="many2many_tags"
+                    readonly="1"
+                />
```
```suggestion
                <field name="pack_putaway_strategy" groups="stock.group_stock_storage_categories" />
                <field name="pack_putaway_sequence" groups="stock.group_stock_storage_categories" />
                <field
                    name="computed_storage_capacity_ids"
                    widget="many2many_tags"
                    readonly="1"
                    groups="stock.group_stock_storage_categories"
                />
```

jbaudoux commented on stock_storage_type/models/stock_location.py:
```diff
@@ -0,0 +1,661 @@
+        "Ordered Children Locations: when a pack is moved to this "
+        "location, a suitable location will be searched in its children "
+        "locations according to the restrictions defined on their "
+        "respective location storage types.",
```
As you added the package type on the product itself also, I would remove the restriction on "Packs"
```suggestion
        string="Put-Away Strategy",
        help="This defines the storage strategy based on package type to use when a product or package is put "
        "away in this location.\n"
        "None: when moved to this location, it will not be put"
        " away any further.\n"
        "Ordered Children Locations: when moved to this "
        "location, a suitable location will be searched in its children "
        "locations according to the restrictions defined on their "
        "respective location storage types.",
```

jbaudoux commented on stock_storage_type/models/stock_location.py:
```diff
@@ -0,0 +1,661 @@
+        "locations according to the restrictions defined on their "
+        "respective location storage types.",
+    )
+    pack_putaway_sequence = fields.Integer()
```
```suggestion
    pack_putaway_sequence = fields.Integer(
        string="Putaway Sequence",
        help="Allow to sort the valid locations by sequence for the storage strategy based on package type")
```
and even rename the field to package_type_putaway_sequence

jbaudoux commented on stock_storage_type/models/stock_storage_category_capacity.py:
```diff
@@ -0,0 +1,96 @@
+
+class StorageCategoryProductCapacity(models.Model):
+
+    _inherit = "stock.storage.category.capacity"
+
+    allow_new_product = fields.Selection(
```
It is missing a _rec_name on allow_new_product.
On the location, it is currently displayed as:
![image](https://user-images.githubusercontent.com/7802725/224349187-65e2cc59-9788-40a7-a593-66e01871a645.png)

rousseldenis replied:
Did this:

https://github.com/OCA/stock-logistics-warehouse/pull/1655

![image](https://user-images.githubusercontent.com/19529533/225358357-703a8fe1-bd5f-4934-a36d-f082254ba55d.png)

jbaudoux commented on stock_storage_type/models/stock_storage_category_capacity.py:
```diff
@@ -0,0 +1,96 @@
+    )
+    # TODO: Check if this is convenient with the constraint on barcode field
+    # in core module
+    active = fields.Boolean(default=True)
```
I would drop this active field

jbaudoux commented on stock_storage_type/models/stock_storage_category.py:
```diff
@@ -0,0 +1,94 @@
+        comodel_name="stock.location", inverse_name="computed_storage_category_id"
+    )
+
+    # TODO: Move these fields in another module ?
```
TODO can be removed

rousseldenis replied:
I've added it during migration as, for me, those fields can be put in a less specialized module.

jbaudoux commented on stock_storage_type/models/stock_location.py:
```diff
@@ -0,0 +1,661 @@
+        putaway_location = super()._putaway_strategy_finalizer(
+            putaway_location, product, quantity, package, packaging, additional_qty
+        )
+        if package:
+            quants = package.quant_ids
+            product = package.single_product_id or product
+        else:
+            quants = self.env["stock.quant"].browse()
+        return self._get_pack_putaway_strategy(putaway_location, quants, product)
+
+    def _get_pack_putaway_strategy(self, putaway_location, quants, product):
```
I would refactor a bit this and rename the method
```suggestion
        return self._get_package_type_putaway_strategy(putaway_location, product, quantity, package)

    def _get_package_type_putaway_strategy(self, putaway_location, product, quantity, package):
        ...
```

rousseldenis replied:
@jbaudoux We need at least the quant as the package_type can come from it

jbaudoux replied:
it is either on the package or on the product, you don't really need the quant

rousseldenis replied:
@jbaudoux In fact, quant is needed further as we need to get some information for it for constraints in the select_allowed_locations(), at least the lot comes from it. So, I will let the function as is.

jbaudoux commented on stock_storage_type/models/stock_location.py:
```diff
@@ -0,0 +1,661 @@
+
+    def _get_pack_putaway_strategy(self, putaway_location, quants, product):
+        package_type = False
+        if quants:
```
```suggestion
        if package.package_type_id:
```

jbaudoux commented on stock_storage_type/models/stock_location.py:
```diff
@@ -0,0 +1,661 @@
+        if quants:
+            package_type = quants.package_id.package_type_id
+            _logger.debug(
+                "Computing putaway for pack %s (%s)"
```
```suggestion
                "Computing putaway for package % of package type %s"
                % (package, package.package_type_id)
```

jbaudoux commented on stock_storage_type/models/stock_location.py:
```diff
@@ -0,0 +1,661 @@
+                "Computing putaway for pack %s (%s)"
+                % (quants.package_id, quants.package_id)
+            )
+        else:
```
```
elif product.package_type_id:
  ....
else:
    return putaway_location

```

jbaudoux commented on stock_storage_type/models/stock_location.py:
```diff
@@ -0,0 +1,661 @@
+            if package_type:
+                _logger.debug(
+                    "Computing putaway for pack %s (%s - from product %s)"
+                    % (quants.package_id, quants.package_id, product)
```
```suggestion
                    "Computing putaway for product %s of package type %s"
                    % (product, product.package_type_id)
```

jbaudoux commented on stock_storage_type/models/product_template.py:
```diff
@@ -0,0 +1,16 @@
+        string="Package type",
+        help="Defines a 'default' package type for this product to be "
+        "applied on packages without product packagings for put-away "
+        "computations.",
```
```suggestion
        help="Defines a 'default' package type for this product to be "
        "applied on packages without product packagings and on put-away computation based on package type for product not in a package.",
```

jbaudoux commented on stock_storage_type/models/stock_quant_package.py:
```diff
@@ -0,0 +1,93 @@
+    @api.model_create_multi
+    def create(self, vals):
+        records = super().create(vals)
+        records._sync_storage_type_from_packaging()
```
Rename in this file storage_type by package_type

rousseldenis commented:
@jbaudoux Stay in touch, I will do the modifications today

rousseldenis committed:
[IMP] stock_storage_type: Use correct domain to see locations from sequences

rousseldenis committed:
[IMP] stock_storage_type: Improve comment and visibility group for categories on locations

rousseldenis committed:
[IMP] stock_storage_type: Improve messages and code logic

rousseldenis committed:
[IMP] stock_storage_type: Change pack_putaway_sequence to package_type_putaway_sequence

rousseldenis committed:
[IMP] stock_storage_type: Better Stock Storage Category name display

rousseldenis committed:
[IMP] stock_storage_type: Remove active field on Stock Storage Category Capacity

rousseldenis commented:
@jbaudoux I think this is ready

jbaudoux reviewed (changes requested):
The README needs to be rewritten.

I would also rename all `pack_putaway_strategy` into `package_type_putaway_strategy`

rousseldenis committed:
[FIX] stock_storage_type: Improve migration scripts

refs #67049

rousseldenis committed:
[IMP] stock_storage_type: Rename putaway strategy method name + allows to apply allowed destinations without packages

Rename the putaway strategy method name to take into account package type

As in some flows, it could have no package defined (no put in pack), fallback on
the product default package type and apply the allowed destination strategy

rousseldenis commented:
@jbaudoux @lmignon This one seems ready and cover now the flow without package (put in pack) for the allowed destinations

rousseldenis committed:
[FIX] stock_storage_type: Fix the behaviour to get the package type at auto assignation from product

rousseldenis committed:
[IMP] stock_storage_type: Set the category from migrated one on locations

rousseldenis committed:
[IMP] stock_storage_type: Add 'Allow new product' on Package Type view

rousseldenis committed:
[IMP] stock_storage_type: Allow to re-apply putaway rules on computed location

rousseldenis committed:
[FIX] stock_storage_type: Remove unused code

rousseldenis committed:
[IMP] stock_storage_type: Take into account complex location views structure

If a view location has a strategy == 'none' and a sequence to be applied, the first call
to _get_putaway_strategy in Odoo core will return the first internal child (and no strategy
will be applied at all).

So, if the location is a view and has no putaway rules for that product, substitute the
child_internal_location_ids field value to self.

jbaudoux reviewed (approved):
LG now

rousseldenis commented:
/ocabot merge nobump
